### PR TITLE
Add file contents cataloger

### LINKED ISSUE #369: Implement selective "retrieve files"
Add file retriever cataloger responsible for retrieving entire file contents for select files:

Configuration:
- Allow for enable/disable
- Takes a scope option as configuration (default: bind to user config/default)
- List of globs that should be persisted (default: [])
- “Max file size”, skipping files over the configured size; default = 1MB

JSON Presentation:
- Adds a new “fileContents” section that is a list of objects, each object should have a file path, a layer, and the base64 encoded file content.

### PR DESCRIPTION
**Pull request by wagoodman**
Enables capturing file content for select files (or glob).

Closes #369

### CODE REVIEW

File: internal/config/application.go [hunk 1]
```diff
@@ -38,6 +38,7 @@ type Application struct {
 	Package           Packages       `yaml:"package" json:"package" mapstructure:"package"`
 	FileMetadata      FileMetadata   `yaml:"file-metadata" json:"file-metadata" mapstructure:"file-metadata"`
 	Secrets           Secrets        `yaml:"secrets" json:"secrets" mapstructure:"secrets"`
+	Contents          contents       `yaml:"contents" json:"contents" mapstructure:"contents"`
```
Comment (wagoodman) [inline internal/config/application.go:41]: Open question @luhring @alfredodeza : I think I should rename this from `contents` to `file-contents` to be consistent with `file-metadata` and another cataloger added in another PR. Any votes / thoughts here?

Comment (luhring) [reply internal/config/application.go:41]: I think that makes sense

Comment (wagoodman) [reply internal/config/application.go]: just made the adjustment

File: syft/file/contents_cataloger.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"fmt"
+	"io"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/source"
+)
+
+type ContentsCataloger struct {
+	globs              []string
+	skipFilesAboveSize int64
```
Comment (luhring) [inline syft/file/contents_cataloger.go:15]: Nit: Clarify unit, e.g. `skipFilesAboveSizeInBytes`

File: syft/file/contents_cataloger.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+func (i *ContentsCataloger) catalogLocation(resolver source.FileResolver, location source.Location) (string, error) {
+	contentReader, err := resolver.FileContentsByLocation(location)
+	if err != nil {
+		return "", err
+	}
+	defer contentReader.Close()
+
+	buf := &bytes.Buffer{}
+	if _, err = io.Copy(base64.NewEncoder(base64.StdEncoding, buf), contentReader); err != nil {
+		return "", fmt.Errorf("unable to observe contents of %+v: %+v", location.RealPath, err)
```
Comment (luhring) [inline syft/file/contents_cataloger.go:64]: Nit: Do we still normally prefer `%w` for errors?

File: syft/file/contents_cataloger_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+func TestContentsCataloger(t *testing.T) {
+	allFiles := []string{"test-fixtures/last/path.txt", "test-fixtures/another-path.txt", "test-fixtures/a-path.txt"}
+
+	tests := []struct {
+		name       string
+		globs      []string
+		maxSize    int64
+		files      []string
+		expected   map[source.Location]string
+		catalogErr bool
```
Comment (luhring) [inline syft/file/contents_cataloger_test.go:19]: `catalogErr` — This doesn't look used. If that's true, I'd remove it from the struct and remove code that uses it in the test body.

File: syft/pkg/cataloger/common/generic_cataloger_test.go [hunk 5]
```diff
@@ -51,7 +51,7 @@ func TestGenericCataloger(t *testing.T) {
 	}
 
 	if len(actualPkgs) != len(expectedPkgs) {
-		t.Fatalf("unexpected packages len: %d", len(actualPkgs))
+		t.Fatalf("unexpected packages len: %d != %d", len(expectedPkgs), len(actualPkgs))
```
Comment (luhring) [inline syft/pkg/cataloger/common/generic_cataloger_test.go:54]: I'd suggest changing this to use an `assert` function from testify to handle this. That should also make it more clear to the consumer of the test output which value is which.

**Review by luhring — CHANGES_REQUESTED** [review]

File: syft/file/contents_cataloger.go (hunk 3 above)
Comment (wagoodman) [reply syft/file/contents_cataloger.go:64]: yup yup, I'll adjust

File: syft/pkg/cataloger/common/generic_cataloger_test.go (hunk 5 above)
Comment (wagoodman) [reply syft/pkg/cataloger/common/generic_cataloger_test.go:54]: Pulled in `assert.Len` to address this 👍

**Review by luhring — APPROVED** [review]

**Comment by wagoodman** [comment]
@luhring rebased to pull in latest changes from main (including JSON schema changes). The largest change was reconciling the schemas into v1.1.0 instead of v1.0.6.

**Pull request merged**